File: tools/music_audio/qwen3_tts/src/nodejs/qwen3_tts-tool.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'
import { NVIDIA_LIBS_PATH, PYTORCH_TORCH_PATH } from '@bridge/constants'

const MODEL_BASE_NAME = 'Qwen3-TTS-12Hz-1.7B-Base'
const MODEL_DESIGN_NAME = 'Qwen3-TTS-12Hz-1.7B-VoiceDesign'
const MODEL_CUSTOM_NAME = 'Qwen3-TTS-12Hz-1.7B-CustomVoice'
const DEFAULT_SETTINGS: Record<string, unknown> = {}
const REQUIRED_SETTINGS: string[] = []

export type SupportedLanguage =
  | 'Auto'
  | 'Chinese'
  | 'English'
  | 'Japanese'
  | 'Korean'
  | 'German'
  | 'French'
  | 'Russian'
  | 'Portuguese'
  | 'Spanish'
  | 'Italian'

interface SynthesizeSpeechTask {
  text: string
  target_language?: SupportedLanguage
  language?: SupportedLanguage
  audio_path?: string
  output_path?: string
  speaker_reference_path?: string
  reference_audio_path?: string
  reference_text?: string
  x_vector_only_mode?: boolean
  max_new_tokens?: number
  do_sample?: boolean
  top_k?: number
  top_p?: number
  temperature?: number
  repetition_penalty?: number
  subtalker_dosample?: boolean
  subtalker_top_k?: number
  subtalker_top_p?: number
  subtalker_temperature?: number
  [key: string]: unknown
}

interface DesignVoiceTask {
  text: string
  target_language?: SupportedLanguage
  language?: SupportedLanguage
  instruct?: string
  audio_path?: string
  output_path?: string
  max_new_tokens?: number
  do_sample?: boolean
  top_k?: number
  top_p?: number
  temperature?: number
  repetition_penalty?: number
  subtalker_dosample?: boolean
  subtalker_top_k?: number
  subtalker_top_p?: number
  subtalker_temperature?: number
  [key: string]: unknown
}

interface CustomVoiceTask {
  text: string
  target_language?: SupportedLanguage
  language?: SupportedLanguage
  /**
   * Vivian for Chinese; Serena for Chinese; Uncle_Fu for Chinese;
   * Dylan for Chinese (Beijing dialect); Eric for Chinese (Sichuan dialect);
   * Ryan for English; Aiden for English; Ono_Anna for Japanese; Sohee for Korean
   */
  speaker:
    | 'Vivian'
    | 'Serena'
    | 'Uncle_Fu'
    | 'Dylan'
    | 'Eric'
    | 'Ryan'
    | 'Aiden'
    | 'Ono_Anna'
    | 'Sohee'
  instruct?: string
  audio_path?: string
  output_path?: string
  max_new_tokens?: number
  do_sample?: boolean
  top_k?: number
  top_p?: number
  temperature?: number
  repetition_penalty?: number
  subtalker_dosample?: boolean
  subtalker_top_k?: number
  subtalker_top_p?: number
  subtalker_temperature?: number
  [key: string]: unknown
}

interface DesignThenSynthesizeTask {
  design_text: string
  design_language?: SupportedLanguage
  design_instruct?: string
  texts: string[]
  languages?: SupportedLanguage[]
  output_paths: string[]
  design_max_new_tokens?: number
  design_do_sample?: boolean
  design_top_k?: number
  design_top_p?: number
  design_temperature?: number
  design_repetition_penalty?: number
  design_subtalker_dosample?: boolean
  design_subtalker_top_k?: number
  design_subtalker_top_p?: number
  design_subtalker_temperature?: number
  max_new_tokens?: number
  do_sample?: boolean
  top_k?: number
  top_p?: number
  temperature?: number
  repetition_penalty?: number
  subtalker_dosample?: boolean
  subtalker_top_k?: number
  subtalker_top_p?: number
  subtalker_temperature?: number
  [key: string]: unknown
}

export default class Qwen3TTSTool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(Qwen3TTSTool.TOOLKIT, this.toolName)
    const toolSettings = ToolkitConfig.loadToolSettings(
      Qwen3TTSTool.TOOLKIT,
      this.toolName,
      DEFAULT_SETTINGS
    )
    this.settings = toolSettings
    this.requiredSettings = REQUIRED_SETTINGS
    this.checkRequiredSettings(this.toolName)
  }

  get toolName(): string {
    // Use the actual config name for toolkit lookup
    return 'qwen3_tts'
  }

  get toolkit(): string {
    return Qwen3TTSTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Synthesize speech with optional voice cloning using Qwen3-TTS.
   */
  async synthesizeSpeech(
    tasks: SynthesizeSpeechTask | SynthesizeSpeechTask[],
    device = 'auto',
    nvidiaLibsPath?: string,
    torchPath?: string
  ): Promise<SynthesizeSpeechTask[]> {
    return this.runTasks(
      'synthesize_speech',
      tasks,
      [MODEL_BASE_NAME],
      device,
      nvidiaLibsPath,
      torchPath
    )
  }

  /**
   * Design a new voice using Qwen3-TTS voice design model.
   */
  async designVoice(
    tasks: DesignVoiceTask | DesignVoiceTask[],
    device = 'auto',
    nvidiaLibsPath?: string,
    torchPath?: string
  ): Promise<DesignVoiceTask[]> {
    return this.runTasks(
      'design_voice',
      tasks,
      [MODEL_DESIGN_NAME],
      device,
      nvidiaLibsPath,
      torchPath
    )
  }

  /**
   * Synthesize speech with a custom voice prompt using Qwen3-TTS.
   */
  async customVoice(
    tasks: CustomVoiceTask | CustomVoiceTask[],
    device = 'auto',
    nvidiaLibsPath?: string,
    torchPath?: string
  ): Promise<CustomVoiceTask[]> {
    return this.runTasks(
      'custom_voice',
      tasks,
      [MODEL_CUSTOM_NAME],
      device,
      nvidiaLibsPath,
      torchPath
    )
  }

  /**
   * Design a voice and then synthesize multiple texts with it.
   */
  async designThenSynthesize(
    tasks: DesignThenSynthesizeTask | DesignThenSynthesizeTask[],
    device = 'auto',
    nvidiaLibsPath?: string,
    torchPath?: string
  ): Promise<DesignThenSynthesizeTask[]> {
    return this.runTasks(
      'design_then_synthesize',
      tasks,
      [MODEL_DESIGN_NAME, MODEL_BASE_NAME],
      device,
      nvidiaLibsPath,
      torchPath
    )
  }

  private async resolveResourceRoot(modelNames: string[]): Promise<string> {
    const modelPaths = await Promise.all(
      modelNames.map((modelName) => this.getResourcePath(modelName))
    )
    const roots = new Set(
      modelPaths.map((modelPath) => path.dirname(modelPath))
    )

    if (roots.size !== 1) {
      throw new Error(
        `Mismatched resource roots for models: ${modelNames.join(', ')}`
      )
    }

    return modelPaths.length > 0 ? path.dirname(modelPaths[0] ?? '') : ''
  }

  private async runTasks<T extends Record<string, unknown>>(
    functionName: string,
    tasks: T | T[],
    modelNames: string[],
    device: string,
    nvidiaLibsPath?: string,
    torchPath?: string
  ): Promise<T[]> {
    const taskArray = Array.isArray(tasks) ? tasks : [tasks]
    let tempDir: string | null = null
    let jsonFilePath: string | null = null

    try {
      const resourceRoot = await this.resolveResourceRoot(modelNames)
      const finalNvidiaLibsPath = nvidiaLibsPath ?? NVIDIA_LIBS_PATH
      const finalTorchPath = torchPath ?? PYTORCH_TORCH_PATH

      tempDir = await fs.promises.mkdtemp(
        path.join(os.tmpdir(), 'qwen3_tts_tasks_')
      )
      jsonFilePath = path.join(tempDir, 'tasks.json')

      await fs.promises.writeFile(
        jsonFilePath,
        JSON.stringify(taskArray, null, 2),
        'utf8'
      )

      const args = [
        '--function',
        functionName,
        '--json_file',
        jsonFilePath,
        '--resource_path',
        resourceRoot,
        '--device',
        device,
        '--torch_path',
        finalTorchPath
      ]

      if (finalNvidiaLibsPath) {
        args.push('--nvidia_libs_path', finalNvidiaLibsPath)
      }

      await this.executeCommand({
        binaryName: 'qwen3_tts',
        args,
        options: { sync: true }
      })

      return taskArray
    } catch (error: unknown) {
      throw new Error(`Qwen3-TTS execution failed: ${(error as Error).message}`)
    }
  }
}

File: tools/music_audio/qwen3_tts/src/python/qwen3_tts_tool.py
import json
import os
import tempfile
from typing import (
    Optional,
    Union,
    List,
    TypedDict,
    TypeVar,
    Mapping,
    Any,
    Sequence,
    Literal,
    cast,
)
from collections.abc import Sequence as SequenceABC, Mapping as MappingABC

from bridges.python.src.sdk.base_tool import BaseTool, ExecuteCommandOptions
from bridges.python.src.sdk.toolkit_config import ToolkitConfig
from bridges.python.src.constants import NVIDIA_LIBS_PATH, PYTORCH_TORCH_PATH

MODEL_BASE_NAME = "Qwen3-TTS-12Hz-1.7B-Base"
MODEL_DESIGN_NAME = "Qwen3-TTS-12Hz-1.7B-VoiceDesign"
MODEL_CUSTOM_NAME = "Qwen3-TTS-12Hz-1.7B-CustomVoice"
DEFAULT_SETTINGS = {}
REQUIRED_SETTINGS = []

SupportedLanguage = Literal[
    "Auto",
    "Chinese",
    "English",
    "Japanese",
    "Korean",
    "German",
    "French",
    "Russian",
    "Portuguese",
    "Spanish",
    "Italian",
]

TTask = TypeVar("TTask", bound=Mapping[str, Any])


class SynthesizeSpeechTask(TypedDict, total=False):
    text: str
    target_language: Optional[SupportedLanguage]
    language: Optional[SupportedLanguage]
    audio_path: Optional[str]
    output_path: Optional[str]
    speaker_reference_path: Optional[str]
    reference_audio_path: Optional[str]
    reference_text: Optional[str]
    x_vector_only_mode: Optional[bool]
    max_new_tokens: Optional[int]
    do_sample: Optional[bool]
    top_k: Optional[int]
    top_p: Optional[float]
    temperature: Optional[float]
    repetition_penalty: Optional[float]
    subtalker_dosample: Optional[bool]
    subtalker_top_k: Optional[int]
    subtalker_top_p: Optional[float]
    subtalker_temperature: Optional[float]


class DesignVoiceTask(TypedDict, total=False):
    text: str
    target_language: Optional[SupportedLanguage]
    language: Optional[SupportedLanguage]
    instruct: Optional[str]
    audio_path: Optional[str]
    output_path: Optional[str]
    max_new_tokens: Optional[int]
    do_sample: Optional[bool]
    top_k: Optional[int]
    top_p: Optional[float]
    temperature: Optional[float]
    repetition_penalty: Optional[float]
    subtalker_dosample: Optional[bool]
    subtalker_top_k: Optional[int]
    subtalker_top_p: Optional[float]
    subtalker_temperature: Optional[float]


class CustomVoiceTask(TypedDict, total=False):
    text: str
    target_language: Optional[SupportedLanguage]
    language: Optional[SupportedLanguage]
    """
    Vivian for Chinese; Serena for Chinese; Uncle_Fu for Chinese;
    Dylan for Chinese (Beijing dialect); Eric for Chinese (Sichuan dialect);
    Ryan for English; Aiden for English; Ono_Anna for Japanese; Sohee for Korean
    """
    speaker: (
        Literal["Vivian"]
        | Literal["Serena"]
        | Literal["Uncle_Fu"]
        | Literal["Dylan"]
        | Literal["Eric"]
        | Literal["Ryan"]
        | Literal["Aiden"]
        | Literal["Ono_Anna"]
        | Literal["Sohee"]
    )
    instruct: Optional[str]
    audio_path: Optional[str]
    output_path: Optional[str]
    max_new_tokens: Optional[int]
    do_sample: Optional[bool]
    top_k: Optional[int]
    top_p: Optional[float]
    temperature: Optional[float]
    repetition_penalty: Optional[float]
    subtalker_dosample: Optional[bool]
    subtalker_top_k: Optional[int]
    subtalker_top_p: Optional[float]
    subtalker_temperature: Optional[float]


class DesignThenSynthesizeTask(TypedDict, total=False):
    design_text: str
    design_language: Optional[SupportedLanguage]
    design_instruct: Optional[str]
    texts: List[str]
    languages: Optional[List[SupportedLanguage]]
    output_paths: List[str]
    design_max_new_tokens: Optional[int]
    design_do_sample: Optional[bool]
    design_top_k: Optional[int]
    design_top_p: Optional[float]
    design_temperature: Optional[float]
    design_repetition_penalty: Optional[float]
    design_subtalker_dosample: Optional[bool]
    design_subtalker_top_k: Optional[int]
    design_subtalker_top_p: Optional[float]
    design_subtalker_temperature: Optional[float]
    max_new_tokens: Optional[int]
    do_sample: Optional[bool]
    top_k: Optional[int]
    top_p: Optional[float]
    temperature: Optional[float]
    repetition_penalty: Optional[float]
    subtalker_dosample: Optional[bool]
    subtalker_top_k: Optional[int]
    subtalker_top_p: Optional[float]
    subtalker_temperature: Optional[float]


class Qwen3TTSTool(BaseTool):
    """
    Tool for text-to-speech, voice cloning, and voice design using Qwen3-TTS.
    """

    TOOLKIT = "music_audio"

    def __init__(self):
        super().__init__()
        # Load configuration from central toolkits directory
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)
        self.settings = ToolkitConfig.load_tool_settings(
            self.TOOLKIT, self.tool_name, DEFAULT_SETTINGS
        )
        self.required_settings = REQUIRED_SETTINGS
        self._check_required_settings(self.tool_name)

    @property
    def tool_name(self) -> str:
        # Use the actual config name for toolkit lookup
        return "qwen3_tts"

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config["description"]

    def synthesize_speech(
        self,
        tasks: Union[SynthesizeSpeechTask, List[SynthesizeSpeechTask]],
        device: str = "auto",
        nvidia_libs_path: Optional[str] = None,
        torch_path: Optional[str] = None,
    ) -> List[SynthesizeSpeechTask]:
        return self._run_tasks(
            "synthesize_speech",
            tasks,
            [MODEL_BASE_NAME],
            device,
            nvidia_libs_path,
            torch_path,
        )

    def design_voice(
        self,
        tasks: Union[DesignVoiceTask, List[DesignVoiceTask]],
        device: str = "auto",
        nvidia_libs_path: Optional[str] = None,
        torch_path: Optional[str] = None,
    ) -> List[DesignVoiceTask]:
        return self._run_tasks(
            "design_voice",
            tasks,
            [MODEL_DESIGN_NAME],
            device,
            nvidia_libs_path,
            torch_path,
        )

    def custom_voice(
        self,
        tasks: Union[CustomVoiceTask, List[CustomVoiceTask]],
        device: str = "auto",
        nvidia_libs_path: Optional[str] = None,
        torch_path: Optional[str] = None,
    ) -> List[CustomVoiceTask]:
        return self._run_tasks(
            "custom_voice",
            tasks,
            [MODEL_CUSTOM_NAME],
            device,
            nvidia_libs_path,
            torch_path,
        )

    def design_then_synthesize(
        self,
        tasks: Union[DesignThenSynthesizeTask, List[DesignThenSynthesizeTask]],
        device: str = "auto",
        nvidia_libs_path: Optional[str] = None,
        torch_path: Optional[str] = None,
    ) -> List[DesignThenSynthesizeTask]:
        return self._run_tasks(
            "design_then_synthesize",
            tasks,
            [MODEL_DESIGN_NAME, MODEL_BASE_NAME],
            device,
            nvidia_libs_path,
            torch_path,
        )

    def _resolve_resource_root(self, model_names: List[str]) -> str:
        model_paths = [self.get_resource_path(model_name) for model_name in model_names]
        roots = {os.path.dirname(model_path) for model_path in model_paths}

        if len(roots) != 1:
            raise Exception(
                f"Mismatched resource roots for models: {', '.join(model_names)}"
            )

        return os.path.dirname(model_paths[0]) if model_paths else ""

    def _run_tasks(
        self,
        function_name: str,
        tasks: Union[TTask, Sequence[TTask]],
        model_names: List[str],
        device: str,
        nvidia_libs_path: Optional[str],
        torch_path: Optional[str],
    ) -> List[TTask]:
        if isinstance(tasks, MappingABC):
            task_list = [cast(TTask, tasks)]
        elif isinstance(tasks, SequenceABC):
            task_list = cast(List[TTask], list(tasks))
        else:
            task_list = [cast(TTask, tasks)]

        try:
            resource_root = self._resolve_resource_root(model_names)
            final_nvidia_libs_path = (
                nvidia_libs_path if nvidia_libs_path is not None else NVIDIA_LIBS_PATH
            )
            final_torch_path = (
                torch_path if torch_path is not None else PYTORCH_TORCH_PATH
            )

            with tempfile.NamedTemporaryFile(
                mode="w", suffix=".json", delete=False, encoding="utf-8"
            ) as temp_file:
                json_file_path = temp_file.name
                json.dump(task_list, temp_file, indent=2, ensure_ascii=False)

            args = [
                "--function",
                function_name,
                "--json_file",
                json_file_path,
                "--resource_path",
                resource_root,
                "--device",
                device,
                "--torch_path",
                final_torch_path,
            ]

            if final_nvidia_libs_path:
                args.extend(["--nvidia_libs_path", final_nvidia_libs_path])

            self.execute_command(
                ExecuteCommandOptions(
                    binary_name="qwen3_tts", args=args, options={"sync": True}
                )
            )

            return task_list
        except Exception as e:
            raise Exception(f"Qwen3-TTS execution failed: {str(e)}")
